File: backend-services/organization/src/labels/labels.service.ts
import { HttpService } from '@nestjs/axios';
import { Injectable } from '@nestjs/common';
import { AxiosError } from 'axios';
import { catchError, firstValueFrom } from 'rxjs';
import { CustomLogger } from '../logger/custom-logger.service';
import { CreateLabelInput } from './dto/create-label.input';
import { UpdateLabelInput } from './dto/update-label.input';
import { Label } from './entities/label.entity';

@Injectable()
export class LabelsService {
  private url = 'http://data-access:3000/labels/';

  constructor(
    private readonly http: HttpService,
    private readonly logger: CustomLogger,
  ) {
    this.logger.setContext('LabelsService');
  }

  async create(createLabelInput: CreateLabelInput): Promise<Label> {
    const { data } = await firstValueFrom(
      this.http.post<Label>(this.url, createLabelInput).pipe(
        catchError((error: AxiosError) => {
          // this.logger.error(error.response.data);
          console.log(error);
          throw 'An error happened!';
        }),
      ),
    );
    return data;
  }

  async findAll(project_id?: number, team_id?: number): Promise<Label[]> {
    const { data } = await firstValueFrom(
      this.http
        .get<Label[]>(this.url, { params: { project_id, team_id } })
        .pipe(
          catchError((error: AxiosError) => {
            this.logger.error(error.response.data);
            throw error.response.data;
          }),
        ),
    );

    return data;
  }

  async findOne(id: number): Promise<Label> {
    const { data } = await firstValueFrom(this.http.get<Label>(this.url + id));
    return data;
  }

  async update(id: number, updateLabelInput: UpdateLabelInput): Promise<Label> {
    const { data } = await firstValueFrom(
      this.http.patch<Label>(`${this.url}${id}`, updateLabelInput),
    );
    return data;
  }

  async remove(id: number): Promise<Label> {
    const { data } = await firstValueFrom(
      this.http.delete<Label>(`${this.url}${id}`),
    );
    return { ...data, id };
  }
}

File: backend-services/projectservice/src/main/java/com/mariusniemet/projectservice/services/LabelsService.java
package com.mariusniemet.projectservice.services;

import com.mariusniemet.projectservice.dto.CreateLabelInput;
import com.mariusniemet.projectservice.dto.UpdateLabelInput;
import com.mariusniemet.projectservice.models.Label;
import com.mariusniemet.projectservice.utils.RestClient;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class LabelsService {
    private final RestClient client;
    private final  String url = "http://localhost:5002/labels/";

    public LabelsService(RestClient client) {
        this.client = client;
    }

    public Label create(CreateLabelInput createLabelInput){
        ResponseEntity<Label> response = this.client.post(url, createLabelInput, Label.class);
        return response.getBody();
    }


    public List<Label> findAll(Integer project_id, Integer team_id) {
        Map<String, String> queryParams = new HashMap<>();

        if(project_id != null){
            queryParams.put("project_id", String.valueOf(project_id));

        }

        if(team_id != null){
            queryParams.put("team_id", String.valueOf(team_id));
        }

        return  this.client.getList(url, queryParams);
    }

    public Label findOne(int id){
        ResponseEntity<Label> response = this.client.get(url+ id, Label.class);
        return response.getBody();
    }

    public Label update(UpdateLabelInput update){
        ResponseEntity<Label> response =  this.client.put(url+ update.getId(), update, Label.class);
        return response.getBody();
    }
}

